docs/specs/instrumentation.md

```diff
@@ -8,34 +8,142 @@ interpreted as described in [RFC 2119].
+catalogue. Metrics and logs SHOULD NOT be emitted directly except for where it
+is otherwise impossible to emit Vector's events, such as in an external crate
+that cannot import Vector's events.
+
+#### Error
+
+An `<Name>Error` event MUST be emitted when an error occurs.
+
+* Properties
+  * `error_code` - An error code for the failure, if applicable.
+    * SHOULD only be specified if it adds additional information beyond
+      `error_type`.
+    * The values for `error_code` for a given error event MUST be a bounded set
+      with relatively low cardinality because it will be used as a metric tag.
+      Examples would be syscall error code. Examples of values that should not
+      be used are raw error messages from `serde` as these are highly variable
+      depending on the input. Instead, these errors should be converted to an
+      error code like `invalid_json`.
+  * `error_type` - The type of error condition. MUST be one of the types listed
+    in the `error_type` enum list in the cue docs.
+  * If any of the above properties are implicit to the specific error
+    type, they MAY be omitted from being represented explicitly in the
+    event fields. However, they MUST still be included in the emitted
+    logs and metrics, as specified below, as if they were present.
+* Metrics
+  * MUST include the defined properties as tags.
+  * MUST increment `<namespace>_errors_total` metric.
+* Logs
+  * MUST log a descriptive, user friendly error message that sufficiently
+    describes the error.
+  * MUST include the defined properties as key-value pairs.
+  * MUST log a message at the `error` level.
+  * SHOULD be rate limited to 10 seconds.
+* Events
+  * MUST emit an [`EventsDropped`] event if the error results in dropping
+    events, or the error event itself MUST meet the `EventsDropped`
+    requirements.
+
+#### EventsDropped
+
+An `<Namespace>EventsDropped` event MUST be emitted when events are dropped.
+If events are dropped due to an error, then the error event should drive the
+emission of this event, meeting the below requirements.
+
+**You MUST NOT emit this event for retriable operations that can recover and
+prevent data loss. For example, a failed HTTP request that will be retried does
+not result in data loss if the retry succeeds.** 
+
+* Properties
+  * `count` - The number of events dropped
+  * `intentional` - Distinguishes if the events were dropped intentionally. For
+    example, events dropped in the `filter` transform are intentionally dropped,
+    while events dropped due to an error in the `remap` transform are
+    unintentionally dropped.
+  * `reason` - A short, user-friendly reason that describes why the events were
+    dropped.
+* Metrics
+  * MUST increment the `<namespace>_discarded_events_total` counter by the
+    number of events discarded.
+  * MUST include the listed properties as tags except the `reason` property.
+* Logs
+  * MUST log a `<count> events <intentional> dropped: <reason>` message.
```

Comment:
I'd recommend making the dynamic parts here attributes, rather than part of the string. This makes it easier to, for example, analyze these messages in downstream systems.

Reply:
Resolved in https://github.com/vectordotdev/vector/pull/12822/commits/d568f96bdabf601dd12760e4466765e9add0e433